language-plutus-core/generators/Language/PlutusCore/Generators/AST.hs

```diff
@@ -1,99 +1,161 @@
 {-# LANGUAGE OverloadedStrings #-}
 
 module Language.PlutusCore.Generators.AST
-    ( genTerm
+    ( AstGen
+    , runAstGen
+    , genTerm
     , genProgram
-    , genVersion
-    , genTyName
-    , genName
-    , genKind
-    , genBuiltinName
-    , genBuiltin
-    , genConstant
-    , genType
-    , simpleRecursive
+    , mangleNames
     ) where
 
+import           PlutusPrelude
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Subst
+import           Language.PlutusCore.Type
+
+import           Control.Lens
+import           Control.Monad.Morph          (hoist)
+import           Control.Monad.Reader
 import qualified Data.ByteString.Lazy         as BSL
+import           Data.Set                     (Set)
+import qualified Data.Set                     as Set
 import           Hedgehog                     hiding (Size, Var)
-import qualified Hedgehog.Gen                 as Gen
+import qualified Hedgehog.Internal.Gen        as Gen
 import qualified Hedgehog.Range               as Range
-import           Language.PlutusCore
-import           Language.PlutusCore.Constant
-import           Language.PlutusCore.Pretty
+
+{- Note [ScopeHandling]
+We intentionally do not distinguish between the type-level and term-level scopes to ensure that
+all the machineries handle variables with same uniques from distinct scopes correctly.
+-}
+
+-- See Note [ScopeHandling].
+type AstGen = GenT (Reader [Name ()])
+
+runAstGen :: MonadGen m => AstGen a -> m a
+runAstGen a = do
+    names <- genNames
+    Gen.liftGen $ hoist (return . flip runReader names) a
 
 genVersion :: MonadGen m => m (Version ())
-genVersion = Version () <$> int' <*> int' <*> int'
-    where int' = Gen.integral_ (Range.linear 0 10)
+genVersion = Version () <$> int' <*> int' <*> int' where
+    int' = Gen.integral_ $ Range.linear 0 10
+
+genNames :: MonadGen m => m [Name ()]
+genNames = do
+    let genUniq = Unique <$> Gen.int (Range.linear 0 100)
+    uniqs <- Set.toList <$> Gen.set (Range.linear 1 20) genUniq
```

Comment:
> we generate a fixed set of names which we will use, of only up to a short size to make it likely that we get reuse.

Thanks for writing down a comment that I can just copy-paste into the source code :)

> where `genFreshName` adds the new name to some state, and `genUsedName` picks from the state?

Sorry, I don't get it. What do you mean? We only generate fresh names in `genNames`.

Reply:
> Sorry, I don't get it. What do you mean? We only generate fresh names in genNames.

Right, this is a replacement for the combination of `genNames` and `genName`. At the moment, `genNames` generates a fixed set of names, and then `genName` picks from it. What I'm suggesting is that rather than generating the names up front, we do it as we go, with the generator choosing either to take a name that's already generated (to make sure we reuse names), or generating a new one (which then gets added to the set of names that can be re-picked).

Not a big deal, though, and this way works.

Reply:
Ah, I see. I think it'd make it harder to control the total amount of names while in the current approach it directly depends on the implicit `hedgehog`'s size parameter. But more importantly, I think your approach creates skewed terms, because `genUsedName` has fewer options in the beginning of generation than in the end of it and so names won't be distributed evenly.

Reply:
Reasonable.